shared/pingpong/accounts.go

```diff
@@ -450,73 +563,136 @@ func prepareApps(accounts map[string]uint64, client libgoal.Client, cfg PpConfig
-		if err != nil {
-			fmt.Printf("Cannot create app txn\n")
-			panic(err)
-		}
+			// generate app program with roughly some number of operations
+			prog := genAppProgram(cfg.AppProgOps, cfg.AppProgHashes, cfg.AppProgHashSize, cfg.AppGlobKeys, cfg.AppLocalKeys)
+			if !cfg.Quiet {
+				fmt.Printf("generated program: \n%s\n", prog)
+			}
 
-		tx, err = client.FillUnsignedTxTemplate(appAccount.Address, 0, 0, cfg.MaxFee, tx)
-		if err != nil {
-			fmt.Printf("Cannot fill app creation txn\n")
-			panic(err)
-		}
+			globSchema := basics.StateSchema{NumByteSlice: 64}
+			locSchema := basics.StateSchema{NumByteSlice: 16}
+			tx, err = client.MakeUnsignedAppCreateTx(transactions.NoOpOC, prog, prog, globSchema, locSchema, nil, nil, nil, nil)
+			if err != nil {
+				fmt.Printf("Cannot create app txn\n")
+				panic(err)
+			}
 
-		// Ensure different txids
-		var note [8]byte
-		crypto.RandBytes(note[:])
-		tx.Note = note[:]
+			tx, err = client.FillUnsignedTxTemplate(appAccount.Address, 0, 0, cfg.MaxFee, tx)
+			if err != nil {
+				fmt.Printf("Cannot fill app creation txn\n")
+				panic(err)
+			}
 
-		signedTxn, signErr := client.SignTransactionWithWallet(h, nil, tx)
-		if signErr != nil {
-			fmt.Printf("Cannot sign app creation txn\n")
-			err = signErr
-			return
+			// Ensure different txids
+			var note [8]byte
+			crypto.RandBytes(note[:])
+			tx.Note = note[:]
+
+			txgroup = append(txgroup, tx)
+			accounts[appAccount.Address] -= tx.Fee.Raw
 		}
 
-		txid, broadcastErr := client.BroadcastTransaction(signedTxn)
-		if broadcastErr != nil {
-			fmt.Printf("Cannot broadcast app creation txn\n")
-			err = broadcastErr
+		err = sendAsGroup(txgroup, client, h)
+		if err != nil {
 			return
 		}
 
 		if !cfg.Quiet {
-			fmt.Printf("Create a new app: txid=%s\n", txid)
+			fmt.Printf("Created new %d apps\n", len(txgroup))
 		}
-
-		accounts[appAccount.Address] -= tx.Fee.Raw
 	}
 
-	var account v1.Account
 	// get these apps
-	for {
-		account, err = client.AccountInformation(appAccount.Address)
-		if err != nil {
-			fmt.Printf("Warning, cannot lookup source account")
-			return
+	var aidxs []uint64
+	appParams = make(map[uint64]v1.AppParams)
+	for _, appAccount := range appAccounts {
+		var account v1.Account
+		for {
+			account, err = client.AccountInformation(appAccount.Address)
+			if err != nil {
+				fmt.Printf("Warning, cannot lookup source account")
+				return
+			}
+			for idx := range account.AppParams {
+				aidxs = append(aidxs, idx)
+			}
+			if len(account.AppParams) >= appsPerAcct || len(aidxs) >= int(cfg.NumApp) {
+				break
+			}
+			time.Sleep(time.Second)
 		}
-		if len(account.AppParams) >= int(cfg.NumApp) {
-			break
+		for k, v := range account.AppParams {
+			appParams[k] = v
+		}
+	}
+
+	// time to opt in to these apps
+	if cfg.NumAppOptIn > 0 {
+		optIns = make(map[uint64][]string)
+		for addr := range accounts {
+			var txgroup []transactions.Transaction
+			addrOptedIn := make(map[uint64]bool)
```

Comment:
consider using rand.Perm, and remove the entry from the indexing array once you're done with it.

Reply:
done